HTMLVideoElement::supportsFullscreen() is incorrect when video uses element fullscreen and the fullscreen API is disabled

pxlcoder opened the pull request:
#### 771504e8eccd7361ee9ef89081333f661e92402c
<pre>
HTMLVideoElement::supportsFullscreen() is incorrect when video uses element fullscreen and the fullscreen API is disabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=249810">https://bugs.webkit.org/show_bug.cgi?id=249810</a>
rdar://103359417

Reviewed by Dean Jackson.

* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::supportsFullscreen const):

Avoid short-circuiting when `ENABLE(VIDEO_USES_ELEMENT_FULLSCREEN)` is true.

Canonical link: <a href="https://commits.webkit.org/258393@main">https://commits.webkit.org/258393@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/11dc01816ec2e86b0b56b9b8fbb1ffa9d26e0502

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101394 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10551 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34452 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110665 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170932 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105374 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11502 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1404 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93788 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108491 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107176 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8744 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91981 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/35272 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90642 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23388 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78269 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4163 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24900 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4218 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1325 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10313 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44387 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5982 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3036 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/html/HTMLVideoElement.cpp:
```diff
@@ -180,7 +180,7 @@ bool HTMLVideoElement::supportsFullscreen(HTMLMediaElementEnums::VideoFullscreen
         return false;
 
-#if PLATFORM(IOS_FAMILY)
+#if PLATFORM(IOS_FAMILY) && !ENABLE(VIDEO_USES_ELEMENT_FULLSCREEN)
```
Do we actually need this ifdef at all? Can we just make the code below that checks all the bits work on iOS?

pxlcoder replied:
Might be possible with clever refactoring, but there's a lot to untangle underneath this code and I'm not seeing an obvious pathway to doing so. We need to avoid running the following on iOS:

```
    if (videoFullscreenMode == HTMLMediaElementEnums::VideoFullscreenModeStandard && !document().settings().fullScreenEnabled())
        return false;
```

`supportsFullScreenForElement` below this code also consults a delegate method in WK1. On WK2, that method consults the bit on settings. Furthermore, both `supportsFullScreenForElement` and `supportsVideoFullscreen` have additional call sites that need to be audited.

We also need to avoid the `!player()->hasVideo()` below `supportsFullScreenForElement` on iOS. 

At best, I think we would just be moving this `#ifdef` into another method.

pxlcoder committed:
HTMLVideoElement::supportsFullscreen() is incorrect when video uses element fullscreen and the fullscreen API is disabled
https://bugs.webkit.org/show_bug.cgi?id=249810
rdar://103359417

Reviewed by Dean Jackson.

* Source/WebCore/html/HTMLVideoElement.cpp:
(WebCore::HTMLVideoElement::supportsFullscreen const):

Avoid short-circuiting when `ENABLE(VIDEO_USES_ELEMENT_FULLSCREEN)` is true.

Canonical link: https://commits.webkit.org/258393@main
